### ISSUE TITLE
connect ECONNREFUSED 0.0.0.0:8080

### ISSUE THREAD

**Issue by a2beckj** [GeoSoftII2020-21/TestRepo#5]
When trying to connect to our backend, which now runs on http://0.0.0.0:8080/api/v1 we get following error: 

> `Can't read well-known document, connecting directly to the specified URL as fallback mechanism. Reason: connect ECONNREFUSED 0.0.0.0:8080`

 (see [Action](https://github.com/GeoSoftII2020-21/TestRepo/runs/1760126369?check_suite_focus=true))

or

> connect ECONNREFUSED 127.0.0.1:80

(see [Action](https://github.com/GeoSoftII2020-21/TestRepo/runs/1760322827?check_suite_focus=true))

We specify that path in our [index.js script](https://github.com/GeoSoftII2020-21/TestRepo/blob/main/.github/actions/index.js).

The containers seem to be reachable within our action, but for some reason they can't be reached within our node-environment in which we call the script... Do you have any advice?

**Comment by nuest** [comment]
I do not know how `.github/actions/index.js` should work - I have only used `.github/workflows/` myself so far.

Can you add some documentation to the README or here as to _what_ you want to achieve?

**Comment by nuest** [comment]
Also, the second error mentions port `80`, not `8080`.

It is also confusing that the actions are all "green" - maybe you can first clarify the actual "thing to test" so that tests fail when they should.

**Comment by m-mohr** [comment]
Node (and other) server tooling often shows it's available on 0.0.0.0, but in fact it's usually available on 127.0.0.1 (or localhost). I guess if you try this IP with the correct port, it should work. Other than that, I fully agree with Daniel.

**Comment by a2beckj** [comment]
Versions we already tried were:

- http://localhost:80/api/v1 bzw. http://127.0.0.1:80/api/v1

- http://localhost:8080/api/v1 bzw. http://127.0.0.1:8080/api/v1

- http://0.0.0.0:80/api/v1

- http://0.0.0.0:8080/api/v1

but none of those solved our problem unfortunately.

@nuest, the test does not fail yet, because we do not have access to any values, that should be validated e.g. number of processes etc... due to the missing connection to the backend. Once this works, we will be able to throw errors/exits. But for testing reasons there should be something like this (in this example with "https://earthengine.openeo.org"):
![image](https://user-images.githubusercontent.com/18323856/105696242-53c6a500-5f03-11eb-918a-684c826f6e3b.png)

These are the replys that we get according to the backend paths:

http://localhost:80/api/v1:
![image](https://user-images.githubusercontent.com/18323856/105698797-ad7c9e80-5f06-11eb-9799-0f6861c6b757.png)

http://localhost:8080/api/v1:
![image](https://user-images.githubusercontent.com/18323856/105698027-9d17f400-5f05-11eb-9e88-dc148b7f35f5.png)

http://0.0.0.0:80/api/v1:
![image](https://user-images.githubusercontent.com/18323856/105699423-a0ac7a80-5f07-11eb-962c-101bcf4a390f.png)

http://0.0.0.0:8080/api/v1:
![image](https://user-images.githubusercontent.com/18323856/105699907-5d9ed700-5f08-11eb-8054-53b1f5124b32.png)

**Comment by m-mohr** [comment]
I can't find an Action where it starts a server. Which one is it? If it's the docker part, you need to ensure that you can actually reach the server that is running inside the container from outside.

A test should also fail if a connection can't be established.

**Comment by a2beckj** [comment]
The index.js is called by this [action](https://github.com/GeoSoftII2020-21/TestRepo/blob/main/.github/actions/action.yml). The docker images are started by this [action](https://github.com/GeoSoftII2020-21/TestRepo/blob/main/.github/workflows/js-client.yml) in step 'Docker run'. In step 'setup node' we setup the server. In step 'check docker containers' we can see that containers are up.

**Comment by nuest** [comment]
Please remove the boilerplate "Example..." in https://github.com/GeoSoftII2020-21/TestRepo/blob/main/.github/actions/action.yml - I stopped considering the file because I though it was not "real" or used.

------

> @nuest, the test does not fail yet, because we do not have access to any values, that should be validated e.g. number of processes etc.

That's not true. When you can successfully use the client against the public API (earthengine.openeo.org) then you can write a test that looks for these values

Not having access to the values is the whole point of writing the test first: write a test that checks if the values exist, and when they eventually exist, the test will switch from fail to success.

------

Again, going through the same steps locally, I get this:

```
$ node .github/actions/index.js 
URL: http://0.0.0.0:8080/api/v1
Client Version: 1.0.2
connect ECONNREFUSED 127.0.0.1:80
```

**Why does it say "refused" on port 80 when I have configured to use port 8080 ?**

When I run the compose file at port 80 and test with port 80, it works:

```
$ node .github/actions/index.js 
URL: http://localhost/api/v1
Client Version: 1.0.2
Server Version: 1.0.0
Number of supported collections: 2
Number of supported processes: 4
```

When I run the compose file at port 80 and test with port 8080 it fails (good) and it also reports the correct failure port (very good):

```
$ node .github/actions/index.js 
URL: http://0.0.0.0:8080/api/v1
Client Version: 1.0.2
Can't read well-known document, connecting directly to the specified URL as fallback mechanism. Reason: connect ECONNREFUSED 0.0.0.0:8080
connect ECONNREFUSED 0.0.0.0:8080
```

So, back to the question above: why does it give an error message on port 80 when I have configured to use 8080, and the endpoint runs at 8080?

```
$ node .github/actions/index.js 
URL: http://localhost:8080/api/v1
Client Version: 1.0.2
connect ECONNREFUSED 127.0.0.1:80
```

**Comment by a2beckj** [comment]
> Not having access to the values is the whole point of writing the test first: write a test that checks if the values exist, and when they eventually exist, the test will switch from fail to success.

In the index.js we do have access by for example filtering the 'processes.processes.length' parameter. But we did not implement that yet. We just wanted to achieve the display of the processes and collections, like you managed to do here: 

![image](https://user-images.githubusercontent.com/18323856/105709666-54b50200-5f16-11eb-8d14-331796df6ee1.png)

And yes, locally we do not encounter any error either. Its just when it comes to starting the images in a github action etc.
Would it be sufficient for you if we provide you with an html, for example, to test the js-client, other than implementing it in a github action?

**Comment by nuest** [comment]
I think you already did a very good job in setting up the tests, including the JS client. My suggestion:

Implement the test for the JS client so that it works locally. I'd say that should be pretty straightforward with a JavaScript testing framework: show that the client can connect to your server implementation and that it finds the expected number of processes etc. Then you create an issue to show how the tests work locally, but no on GitHub. Then you can disable the JS workflow and in your report a reference to the issue. Then it's a problem with GH Action only, but not a shortcoming on your testing at all.

_Einverstanden?_

**One last question:** Have you tried running `index.js` directory from the workflow file? Maybe the action is running in some other environment?

**Comment by a2beckj** [comment]
> One last question: Have you tried running index.js directory from the workflow file? Maybe the action is running in some other environment?

Yes we tried with various attempts, but the only working solution we found was [this one](https://dev.to/talk2megooseman/talking-github-actions-with-benjamin-lannon-aka-lannonbr-1acf) or [this one](https://docs.github.com/en/actions/creating-actions/creating-a-javascript-action). So just by creating a seperate action with the only purpose to run another script.

We'd be _einverstanden_ with your suggestion mentioned before. So just creating an issue like this one, to show what to do in order to make it run locally, right?
